src/codemirror/model.ts

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import * as CodeMirror
+  from 'codemirror';
+
+import {
+  ISignal, clearSignalData, defineSignal
+} from 'phosphor/lib/core/signaling';
+
+import {
+  CodeEditor
+} from '../codeeditor/editor';
+
+import {
+  IChangedArgs
+} from '../common/interfaces';
+
+import {
+  ObservableVector
+} from '../common/observablevector';
+
+
+/**
+ * An implementation of the code editor model using code mirror.
+ */
+export
+class CodeMirrorModel implements CodeEditor.IModel {
+  /**
+   * Construct a new codemirror model.
+   */
+  constructor() {
+    let doc = this._doc = new CodeMirror.Doc('');
```

Comment:
Do you want to create a CodeMirror instance and swap its the Doc? Otherwise, would it make sense to pass the Doc as parameter?